### PR TITLE
Implementation of recoil estimate into fit and fit interface additions

### PR DESCRIPTION
**Pull request by azotz**
- working version of the fit with transverse recoil estimate via MET
- added flag in GEF to use pt corrections
- added RotationSignificance to TPTRObject
- added information about fit solution with worse chi2 in GEFObject
- removed unused functions in DiTauConstrainedFitter which were only used during development

### CODE REVIEW

File: FitSoftware/interface/GlobalEventFit.h [hunk 1]
```diff
@@ -58,9 +59,11 @@ class GlobalEventFit{
 	  MassConstraint_ = MassConstraint;
 	  useDefaultMassConstraint_ = false;
 	}
+	void SetCorrectPt(bool correct){correctPt_ = correct;}
```
Comment (bkargoll) [inline FitSoftware/interface/GlobalEventFit.h]: @azotz : Could you document somewhere what correctPt_ is and how it should be used? It seems it is meant to be set in the analysis code, i.e. in AnalysisTools. Thus, a hint would be good here.

File: FitSoftware/src/DiTauConstrainedFitter.cc [hunk 2]
```diff
@@ -548,6 +614,7 @@ DiTauConstrainedFitter::EstimateTauDirectionAdvanced(TMatrixT<double> &inpar){
   TVector3 TauMuDir = PointGuess - PV;
 
   outpar(0,0) = TauMuDir.Theta();
+  //outpar(1,0) = TauMuDir.Phi();
```
Comment (bkargoll) [inline FitSoftware/src/DiTauConstrainedFitter.cc]: Remove?

File: FitSoftware/src/DiTauConstrainedFitter.cc [hunk 3]
```diff
@@ -577,11 +644,22 @@ DiTauConstrainedFitter::EstimateTauKinematic(TMatrixT<double> &inpar){
...
-  outpar(0,0) = TauMuP*cos(inpar(1,0))*sin(inpar(0,0));
-  outpar(1,0) = TauMuP*sin(inpar(1,0))*sin(inpar(0,0));
-  outpar(2,0) = TauMuP*cos(inpar(0,0));
+  outpar(0,0) = P_TauMu*cos(inpar(1,0))*sin(inpar(0,0));
+  outpar(1,0) = P_TauMu*sin(inpar(1,0))*sin(inpar(0,0));
+  outpar(2,0) = P_TauMu*cos(inpar(0,0));
+
+  //outpar(0,0) = TauMuP*cos(inpar(1,0))*sin(inpar(0,0));
+  //outpar(1,0) = TauMuP*sin(inpar(1,0))*sin(inpar(0,0));
+  //outpar(2,0) = TauMuP*cos(inpar(0,0));
```
Comment (bkargoll) [inline FitSoftware/src/DiTauConstrainedFitter.cc]: Remove?

File: FitSoftware/src/TPTRObject.cc [hunk 4]
```diff
@@ -11,13 +11,14 @@ TPTRObject::TPTRObject(){
...
 
-TPTRObject::TPTRObject(LorentzVectorParticle A1, std::vector<LorentzVectorParticle> Taus, std::vector<LorentzVectorParticle> Neutrinos, bool isambiguous, bool isvalid){
+TPTRObject::TPTRObject(LorentzVectorParticle A1, std::vector<LorentzVectorParticle> Taus, std::vector<LorentzVectorParticle> Neutrinos, bool isambiguous, double RotSig, bool isvalid){
   isvalid_ = isvalid;
   isambiguous_ = isambiguous;
   if(isvalid_){
 	A1_ = A1;
 	Taus_ = Taus;
 	Neutrinos_ = Neutrinos;
+	if(isambiguous_) RotationSignificance_ = RotSig;
```
Comment (bkargoll) [inline FitSoftware/src/TPTRObject.cc]: If the solution is not ambiguous, the user should not have to provide a rotation significance, but he/she has to do so at the moment. Either provide a default value for RotSig, or provide a second (i.e. overloaded) constructor without isambiguous and RotSig, which sets these values to false and 0, respectively.

**Comment by bkargoll** [comment]
Only small comments from my side. I propose to fix them on Monday and merge then.

File: FitSoftware/src/TPTRObject.cc (hunk 4 above)
Comment (azotz) [inline FitSoftware/src/TPTRObject.cc]: Actually this is a bug. There is no rotation if there is an ambiguity. I will fix that and also give it a default value if there is an ambiguity.

File: FitSoftware/src/DiTauConstrainedFitter.cc (hunk 3 above)
Comment (azotz) [inline FitSoftware/src/DiTauConstrainedFitter.cc]: Will remove.

File: FitSoftware/src/DiTauConstrainedFitter.cc (hunk 2 above)
Comment (azotz) [inline FitSoftware/src/DiTauConstrainedFitter.cc]: Yes

**Comment by bkargoll** [comment]
Thanks for the cleanup, @azotz. It seems it was worth it.
Please notify the framework mailing list.

**Pull request merged**